### PR TITLE
OdhApplication, OdhDocument are k8 CRDs

### PR DESCRIPTION
**Pull request by kkasravi**
- [x] updated frontend/src, backend/src to fetch/watch OdhApplication, OdhDocument k8 resources
- [x] moved data/{applications,documents} to install/deploy/base/{odh-application.yaml,odh-document.yaml} 
- [x] added CRDs for OdhApplication, OdhDocument to install/deploy/base
- [x] updated Makefile targets deploy, build, build-local, undeploy, defaults

### CODE REVIEW

**Comment by kkasravi** [comment]
@jeff-phillips-18 

Hi Jeff - this is the PR from red-hat-data-services/odh-dashboard.
Could I ask you to approve the workflow? Please review when you get a chance.
Thanks.

**Comment by jeff-phillips-18** [comment]
@kkasravi Thanks! I will take a look shortly. Will the resources (apps and docs) will be added upon deployment or is there some action necessary to add them?

@cfchase Can you also take a look?

**Comment by kkasravi** [comment]
yes 'make deploy' will deploy the apps and docs to the cluster.

I need to open a sibling PR in odh-manifests so that odh-manifests/odh-dashboard can be deployed in the same way (but with the auth overlay).

**Comment by cfchase** [comment]
This is a really good direction to head in so we can dynamically add applications, etc.  In fact I believe it's necessary for some other requirements. In my opinion, though, all the CRD and deployment work needs to go into the applicable operator repos and should not be in the odh-dashboard.  For that reason, I think we need to get @vpavlin, @nakfour, and @crobby involved to plan the CRDs and their deployment.  This work would be split up and separate the responsibilities.

**Comment by kkasravi** [comment]
> This is a really good direction to head in so we can dynamically add applications, etc. In fact I believe it's necessary for some other requirements. In my opinion, though, all the CRD and deployment work needs to go into the applicable operator repos and should not be in the odh-dashboard. For that reason, I think we need to get @vpavlin, @nakfour, and @crobby involved to plan the CRDs and their deployment. This work would be split up and separate the responsibilities.

I agree. I'm in the middle of creating an aikit-operator that adds an ImageStream (for JupyterHub) and several OdhDocuments when its run.

I had a few questions related to OdhApplications that are 'Enabled' versus those that appear under Explore 
- Do all apps in 'Explore' move to 'Enabled' once their action is executed? eg if an Explore App has you run an operator that adds a new container to the jupyterhub spawner dropdown menu and adds new document types does it need to be shown as an enabled app afterwards?

**Comment by jeff-phillips-18** [comment]
> * Do all apps in 'Explore' move to 'Enabled' once their action is executed? eg if an Explore App has you run an operator that adds a new container to the jupyterhub spawner dropdown menu and adds new document types does it need to be shown as an enabled app afterwards?

Once the app is discovered to be enabled, it will appear in the `Enabled` tab. There are various ways of enabling (based on the `enabled` settings in the yaml file for the app.

File: install/deploy/base/role.yaml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  verbs:
+  - get
+  - patch
+  - update
+- apiGroups:
+  - documents.console.openshift.io
+  resources:
+  - odhdocuments
+  verbs:
+  - create
```
Comment (vpavlin) [inline install/deploy/base/role.yaml:85]: Why does the dashboard need to create/edit the documents? Should it be enough to get/list/watch them?

**Comment by vpavlin** [comment]
I like the direction as well.

It feels like the CRDs and actual manifests would fit into `odh-manifests` the best, but that would need some tweaking in the `Makefile` to simplify development.

I cannot really comment on the TypeScript code, but the yamls look reasonable at a first glance.

File: install/deploy/base/role.yaml (hunk 1 above)
Comment (kkasravi) [reply install/deploy/base/role.yaml:85]: @vpavlin makes sense. I'll update the role.

**Comment by kkasravi** [comment]
@vpavlin 

> I like the direction as well.
> 
> It feels like the CRDs and actual manifests would fit into `odh-manifests` the best, but that would need some tweaking in the `Makefile` to simplify development.
> 
> I cannot really comment on the TypeScript code, but the yamls look reasonable at a first glance.

thanks for the review. I've modified odh-manifests already along these lines. I'll open the PR in odh-manifests within the red-hat-data-services org. 

@vpavlin @jeff-phillips-18 

One aspect of this PR is I've only done OdhApplication, OdhDocument. The directory ./data/{getting-started,quickstarts} should also be converted to and watched as k8 kinds rather than being read from disk. I think these can be done as follow PRs.

Note: the backend doesn't act like a typical k8 operator/controller despite watching for OdhApplication, OdhDocument resources. As an operator/controller, it should update OdhApplication, OdhDocument status, add itself as the owner and add a finalizer. I think this is doable in typescript using @kubernetes/client-node though i've typically done it in golang. This can also be done as a follow up PR. 

I need to rebase as well and address @jeff-phillips-18 comments. 

thanks again
Kam

**Comment by kkasravi** [comment]
> > * Do all apps in 'Explore' move to 'Enabled' once their action is executed? eg if an Explore App has you run an operator that adds a new container to the jupyterhub spawner dropdown menu and adds new document types does it need to be shown as an enabled app afterwards?
> 
> Once the app is discovered to be enabled, it will appear in the `Enabled` tab. There are various ways of enabling (based on the `enabled` settings in the yaml file for the app).

thanks @jeff-phillips-18 for this feedback. The 'enabled' settings would be [these](https://github.com/opendatahub-io/odh-dashboard/blob/master/backend/src/types.ts#L125)?

**Comment by kkasravi** [comment]
@vpavlin @jeff-phillips-18 could I ask one of you approve the PR [workflow](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)? thanks.

**Comment by jeff-phillips-18** [comment]
> The 'enabled' settings would be [these](https://github.com/opendatahub-io/odh-dashboard/blob/master/backend/src/types.ts#L125)?

correct

**Comment by jeff-phillips-18** [comment]
> @vpavlin @jeff-phillips-18 could I ask one of you approve the PR [workflow](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)? thanks.

done

**Comment by kkasravi** [comment]
@jeff-phillips-18 @vpavlin 

I've updated this branch to sync with whats currently in data/{applications,docs}. EG:
```
data/applications/* -> install/deploy/base/odh-applications.yaml
data/docs/* -> install/deploy/base/odh-documents.yaml
```

Please have another look. I can continue converting the remaining files under "./data/ to be k8 kinds once this is merged. 

thanks.

**Comment by jeff-phillips-18** [comment]
Closing in favor of https://github.com/opendatahub-io/odh-manifests/pull/482 and https://github.com/opendatahub-io/odh-dashboard/pull/97

**Pull request closed without merging**